op-challenger/game/fault/preimages/large.go

```diff
@@ -31,24 +52,72 @@ func NewLargePreimageUploader(logger log.Logger, txMgr txmgr.TxManager, contract
 	return &LargePreimageUploader{logger, txMgr, contract}
 }
 
+// GetKeccakLeaf returns the keccak leaf for the preimage data at the specified offset.
+func GetKeccakLeaf(data *types.PreimageOracleData, offset uint32) []byte {
+	offset = offset * matrix.LeafSize
+	end := offset + matrix.LeafSize
+	if offset >= uint32(len(data.OracleData)) {
+		return make([]byte, matrix.LeafSize)
+	}
+	if end > uint32(len(data.OracleData)) {
+		padded := make([]byte, matrix.LeafSize)
+		data := data.OracleData[offset:]
+		copy(padded[matrix.LeafSize-len(data):], data)
+		return padded
+	}
+	return data.OracleData[offset:end]
+}
+
 func (p *LargePreimageUploader) UploadPreimage(ctx context.Context, parent uint64, data *types.PreimageOracleData) error {
-	// todo(proofs#467): generate the full preimage
-	// todo(proofs#467): run the preimage through the keccak permutation, hashing
-	//                   the intermediate state matrix after each block is applied.
-	// todo(proofs#467): split up the preimage into chunks and submit the preimages
-	//                   and state commitments to the preimage oracle contract using
-	//                   `PreimageOracle.addLeavesLPP` (`_finalize` = false).
+	// Split the preimage data into chunks of size [MaxChunkSize] (except the last chunk).
+	stateMatrix := matrix.NewStateMatrix()
+	chunk := make([]byte, 0, MaxChunkSize)
+	calls := []Chunk{}
+	commitments := make([][32]byte, 0, MaxLeafsPerChunk)
+	in := bytes.NewReader(data.OracleData)
+	for i := 0; ; i++ {
+		// Absorb the next preimage chunk leaf and run the keccak permutation.
+		err := stateMatrix.AbsorbNextLeaf(in)
+		chunk = append(chunk, GetKeccakLeaf(data, uint32(i))...)
+		commitments = append(commitments, stateMatrix.StateCommitment())
+		if errors.Is(err, io.EOF) {
+			calls = append(calls, Chunk{chunk, commitments[:], true})
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed to absorb leaf: %w", err)
+		}
+
+		// Only create a call if the chunk is full.
+		if len(chunk) >= MaxChunkSize {
+			calls = append(calls, Chunk{chunk, commitments[:], false})
+			chunk = make([]byte, 0, MaxChunkSize)
+			commitments = make([][32]byte, 0, MaxLeafsPerChunk)
+		}
+	}
+	// The last call **must** be finalized.
+	if len(calls) > 0 {
+		calls[len(calls)-1].Finalize = true
+	}
```

Comment:
This should already happen because we append a chunk with finalise = true when we get an EOF.  If the last call didn't come from that case, then we probably didn't actually submit all the data.